eips_code/eipXXXX.py

```diff
@@ -0,0 +1,96 @@
+from dataclasses import dataclass
+from eipXXXX_table import TABLE, OP_RJUMP, OP_RJUMPI, OP_CALLF, OP_RETF
+from eip3670 import validate_code, immediate_sizes, valid_opcodes, terminating_opcodes
+
+valid_opcodes += [OP_RJUMP, OP_RJUMPI, OP_CALLF, OP_RETF]
+immediate_sizes[OP_RJUMP] = 2
+immediate_sizes[OP_RJUMPI] = 2
+immediate_sizes[OP_CALLF] = 2
+terminating_opcodes += [OP_RETF]
+
+
+@dataclass
+class FunctionType:
+    inputs: int
+    outputs: int
+
+
+def validate_function(func_id: int, code: bytes, types: list[FunctionType] = [FunctionType(0, 0)]) -> int:
+    assert func_id >= 0
+    assert types[func_id].inputs >= 0
+    assert types[func_id].outputs >= 0
+
+    validate_code(code)
+
+    stack_heights = {}
+    start_stack_height = types[func_id].inputs
+    max_stack_height = start_stack_height
+
+    # queue of instructions to analyze, list of (pos, stack_height) pairs
+    worklist = [(0, start_stack_height)]
+
+    while worklist:
+        pos, stack_height = worklist.pop(0)
+        while True:
+            assert pos < len(code), "code is invalid"
+            op = code[pos]
+            info = TABLE[op]
+
+            # Detect stack underflow
+            if stack_height < info.stack_height_required:
+                return -101
+
+            # Check if stack height (type arity) at given position is the same
+            # for all control flow paths reaching this position.
+            if pos in stack_heights:
+                if stack_height != stack_heights[pos]:
+                    return -102
+                else:
+                    break
+            else:
+                stack_heights[pos] = stack_height
+
+            stack_height += info.stack_height_change
+
+            if stack_height >= 1023:
+                return -105
+
+            max_stack_height = max(max_stack_height, stack_height)
+
+            # Handle jumps
+            if op == OP_RJUMP:
+                offset = int.from_bytes(code[pos + 1:pos + 3], byteorder="big", signed=True)
+                pos += 3 + offset  # pos is valid for validated code.
+
+            elif op == OP_RJUMPI:
+                offset = int.from_bytes(code[pos + 1:pos + 3], byteorder="big", signed=True)
+                # Save True branch for later and continue to False branch.
+                # TODO: If target is already visited, we can omit adding to the worklist.
```

Comment:
We could, but then we need to duplicate the check for `stack_height != stack_heights[pos]`, seems simpler code to just add to worklist.